### PR TITLE
TLS 1.3 code added to ssl_client2.c and ssl_server2.c

### PR DESCRIPTION
**Pull request by hannestschofenig**

### CODE REVIEW

File: programs/ssl/ssl_client2.c [hunk 1]
```diff
@@ -217,10 +216,19 @@ int main( void )
...
+    "    psk_identity=%%s         default: \"Client_identity\"\n" \
+    "    key_exchange_modes=%%s   default: all\n"     \
+    "                             options: psk, psk_dhe, ecdhe_ecdsa, psk_all, all\n" 
+#else 
 #define USAGE_PSK_RAW                                               \
     "    psk=%%s              default: \"\" (disabled)\n"     \
     "                          The PSK values are in hex, without 0x.\n" \
-    "    psk_identity=%%s     default: \"Client_identity\"\n"
+    "    psk_identity=%%s     default: \"Client_identity\"\n" 
+#endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: The existing documentation for `psk` and `psk_identity` can be retained - we only need to add the documentation of `key_exchange_modes` conditional on `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL`. 
```c
#define USAGE_PSK_RAW                                               \
    "    psk=%%s                  default: \"\" (disabled)\n"       \
    "                             The PSK values are in hex, without 0x.\n" \
    "    psk_identity=%%s         default: \"Client_identity\"\n"
#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
#define USAGE_KEX_MODES \
    "    key_exchange_modes=%%s   default: all\n"     \
    "                             options: psk, psk_dhe, ecdhe_ecdsa, psk_all, all\n" 
#else 
#define USAGE_KEX_MODES 
#endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
```
and then `USAGE_KEX_MODES` needs to be added somewhere below.

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -390,6 +398,40 @@ int main( void )
 #define USAGE_SERIALIZATION ""
 #endif
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: The existing code includes the protocol version identifiers regardless of whether the respective version is enabled in the config. I'd suggest to continue that pattern and unconditionally add `tls1_3` and `dtls1_3` for the documentation of possible options.

File: programs/ssl/ssl_client2.c (hunk 1 above)
Comment (hanno-becker) [reply programs/ssl/ssl_client2.c]: This would go well next to `USAGE_EARLY_DATA` and friends below, which follow that pattern.

File: programs/ssl/ssl_client2.c [hunk 3]
```diff
@@ -450,15 +492,17 @@ int main( void )
     USAGE_REPRODUCIBLE                                      \
     USAGE_CURVES                                            \
     USAGE_RECSPLIT                                          \
+    USAGE_EARLY_DATA                                        \
+    USAGE_NAMED_GROUP                                       \
+    USAGE_KEY_SHARE_NAMED_GROUPS                            \
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:497]: Here we could also add `USAGE_KEX_MODES`.

File: programs/ssl/ssl_client2.c [hunk 4]
```diff
@@ -549,6 +594,14 @@ struct options
                                  * after renegotiation                      */
     int reproducible;           /* make communication reproducible          */
     int skip_close_notify;      /* skip sending the close_notify alert      */
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+    const char *named_groups_string;    /* list of named groups             */
+    const char *key_share_named_groups_string;    /* list of named groups   */
+    int key_exchange_modes;     /* supported key exchange modes  */
+#if defined(MBEDTLS_ZERO_RTT)
+    int early_data;             /* support for early data */
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Minor: Keep the alignment:
```suggestion
    const char *named_groups_string;    /* list of named groups             */
    const char *key_share_named_groups_string;    /* list of named groups   */
    int key_exchange_modes;     /* supported key exchange modes             */
#if defined(MBEDTLS_ZERO_RTT)
    int early_data;             /* support for early data                   */
```

File: programs/ssl/ssl_client2.c [hunk 5]
```diff
@@ -1158,7 +1211,27 @@ int main( int argc, char *argv[] )
     mbedtls_ctr_drbg_context ctr_drbg;
     mbedtls_ssl_context ssl;
     mbedtls_ssl_config conf;
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+
+#if defined(MBEDTLS_ECP_C)
+    mbedtls_ecp_group_id named_groups_list[MAX_NAMED_GROUPS]; /* list of named groups */
+    mbedtls_ecp_group_id key_share_named_groups_list[MAX_NAMED_GROUPS]; /* list of named groups for key share*/
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Minor: Overly long line.

File: programs/ssl/ssl_client2.c [hunk 6]
```diff
@@ -1158,7 +1211,27 @@ int main( int argc, char *argv[] )
...
+#if defined(MBEDTLS_ZERO_RTT)
+    char early_data[] = "early data test";
+#endif /* MBEDTLS_ZERO_RTT */
+
+#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET)
+    mbedtls_ssl_ticket ticket;
+#endif /* MBEDTLS_SSL_NEW_SESSION_TICKET */
+
+#else 
     mbedtls_ssl_session saved_session;
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:1228]: IIUC this should be enabled even if `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` is set, because ultimately we want the library to build and be usable with both TLS 1.2 and TLS 1.3 enabled.

File: programs/ssl/ssl_client2.c [hunk 7]
```diff
@@ -1196,7 +1269,22 @@ int main( int argc, char *argv[] )
...
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+
+#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET)
+    mbedtls_ssl_init_client_ticket( &ticket );
+#endif /* MBEDTLS_SSL_NEW_SESSION_TICKET */
+
+#if defined(MBEDTLS_ECP_C)
+    memset( (void *) named_groups_list, MBEDTLS_ECP_DP_NONE, sizeof( named_groups_list ) );
+    memset( (void *) key_share_named_groups_list, MBEDTLS_ECP_DP_NONE, sizeof( key_share_named_groups_list ) );
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Minor: Overly long line.

File: programs/ssl/ssl_client2.c [hunk 8]
```diff
@@ -1499,7 +1593,7 @@ int main( int argc, char *argv[] )
         else if( strcmp( p, "tickets" ) == 0 )
         {
             opt.tickets = atoi( q );
-            if( opt.tickets < 0 || opt.tickets > 2 )
+            if( opt.tickets < 0 || opt.tickets > 1 )
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:1589]: What's the rationale for this change?

File: programs/ssl/ssl_client2.c [hunk 9]
```diff
@@ -1539,6 +1633,41 @@ int main( int argc, char *argv[] )
                 default: goto usage;
             }
         }
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+
+#if defined(MBEDTLS_ZERO_RTT)
+        else if( strcmp( p, "early_data" ) == 0 )
+        {
+            if( strcmp( q, "disabled" ) == 0 )
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Binary command line flags in `ssl_{client|server}2` are usually enabled/disabled with `0` or `1` instead of `enabled/disabled`.

File: programs/ssl/ssl_client2.c [hunk 10]
```diff
@@ -1890,7 +2044,7 @@ int main( int argc, char *argv[] )
     }
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
 
-#if defined(MBEDTLS_ECP_C)
+#if defined(MBEDTLS_ECP_C) && !defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Why was this guard added? I understand that the code that's now guarded is TLS 1.2 specific, but we want the code to build with both TLS 1.2 and TLS 1.3 enabled.

File: programs/ssl/ssl_client2.c [hunk 11]
```diff
@@ -2352,12 +2598,20 @@ int main( int argc, char *argv[] )
...
 
 #if defined(MBEDTLS_ECP_C)
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+    if( named_groups_list[0] != MBEDTLS_ECP_DP_NONE )
+        mbedtls_ssl_conf_curves(&conf, named_groups_list);
+
+    if( key_share_named_groups_list[0] != MBEDTLS_ECP_DP_NONE )
+        mbedtls_ssl_conf_key_share_curves( &conf, key_share_named_groups_list );
+#else 
     if( opt.curves != NULL &&
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:2606]: Same here: If I understand correctly, there should be a runtime check deciding which configuration API to call, not a compile-time check.

File: programs/ssl/ssl_client2.c [hunk 12]
```diff
@@ -2526,14 +2780,33 @@ int main( int argc, char *argv[] )
         }
     }
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Same here: The check should happen at runtime, not compile-time.

**Review by hanno-becker — CHANGES_REQUESTED** [review]
@hannestschofenig Thanks a lot for working on consolidating the example programs again :+1: 

Overall it looks good to me, but my main request for rework is that currently there are a lot of compile-time 'branches' distinguishing between TLS 1.3 and TLS 1.2 which should be runtime checks. See the comments for details.

File: programs/ssl/ssl_client2.c (hunk 8 above)
Comment (hannestschofenig) [reply programs/ssl/ssl_client2.c:1589]: There is only the option to enable and disable the use of tickets for the ssl_client2.c
